### ISSUE TITLE
AVX2 detection (Kubuntu 19.10 virtual machine inside VMWare)

### ISSUE THREAD

**Issue by ad8e** [servo/pathfinder#238]
I ran:
>sudo apt install libgtk-3-dev libsdl2-dev
>cd demo/native
>RUSTFLAGS="-C target-cpu=native" cargo run --release --bin demo

I found that libgtk-3-dev was necessary. Also, I added "--bin demo" because the instructions from README.md result in "error: `cargo run` could not determine which binary to run. Use the `--bin` option to specify a binary, or the `default-run` manifest key."

The program started, then crashed. GDB output:

>Starting program: /home/ad/Downloads/pathfinder-master/target/release/demo
...
>Thread 2 "demo" received signal SIGILL, Illegal instruction.
(gdb) bt
#0  0x00005555557db434 in core::core_arch::x86::avx2::_mm_srlv_epi32 ()
#1  0x00005555557db56c in _ZN19pathfinder_renderer7builder60_$LT$impl$u20$pathfinder_renderer..gpu_data..BuiltObject$GT$8add_fill17hb5cc38b7672cae62E.llvm.14562274008100250132 ()
#2  0x00005555557db7e6 in pathfinder_renderer::builder::<impl pathfinder_renderer::gpu_data::BuiltObject>::generate_fill_primitives_for_line ()
#3  0x00005555557daef0 in pathfinder_renderer::tiles::ActiveEdge::process ()
#4  0x00005555557da7f0 in pathfinder_renderer::tiles::process_active_segment ()
#5  0x00005555557da1c2 in pathfinder_renderer::tiles::Tiler::generate_tiles ()
#6  0x00005555557d6dab in pathfinder_renderer::builder::SceneBuilder::build_path ()
#7  0x00005555555afa58 in rayon::iter::plumbing::Producer::fold_with ()
#8  0x00005555555d6c28 in rayon::iter::plumbing::bridge_producer_consumer::helper ()

My system doesn't support AVX2. It's an Ivy Bridge mobile CPU. If detection is not yet implemented, then don't worry about it. If detection is already implemented, it may be the VM interfering.

**Comment by quornian** [comment]
I'm seeing the same. Another non-AVX2 CPU, Intel(R) Xeon(R) X5660.

If I check out the commit before this one, it works for me:
https://github.com/servo/pathfinder/commit/b886c157c18b52881013ac029642450cc07e9bb9

**Comment by quornian** [comment]
Alternatively, the shift that causes this can be swapped for a slower, non-SIMD version of the same, with this patch:
```diff
diff --git a/renderer/src/builder.rs b/renderer/src/builder.rs
index 9e53c49..732e362 100644
--- a/renderer/src/builder.rs
+++ b/renderer/src/builder.rs
@@ -192,13 +192,15 @@ impl BuiltObject {
         let alpha_tile_index = self.get_or_allocate_alpha_tile_index(builder, tile_coords);
 
         // Pack whole pixels.
-        let mut px = (segment & I32x4::splat(0xf00)) >> I32x4::new(8, 4, 8, 4);
-        px = px | px.yxwz();
+        let px = segment & I32x4::splat(0xf00);
 
         // Pack instance data.
         debug!("... OK, pushing");
         self.fills.push(FillBatchPrimitive {
-            px: LineSegmentU4 { from: px[0] as u8, to: px[2] as u8 },
+            px: LineSegmentU4 {
+                from: (px[0] >> 8) as u8 | (px[1] >> 4) as u8,
+                to: (px[2] >> 8) as u8 | (px[3] >> 4) as u8
+            },
             subpx: LineSegmentU8 {
                 from_x: from_x as u8,
                 from_y: from_y as u8,

```
 (applicable to master's latest, which at time of writing is 5d5d30346bf4703ba49244e85bcb8c6ed7df48e9)

I'm not too familiar with SIMD or feature detection, otherwise I'd offer a more complete solution as a PR.

**Comment by linkmauve** [comment]
This issue has been fixed in 0bb6b8885973c7755be6d5fc5900314c6df4fb3a, you can close it.
